site/content/blog/2020-06-04-svelte-and-typescript.md

```diff
@@ -0,0 +1,109 @@
+
+#### How does it work?
+
+To understand the two main parts of TypeScript support, we'll compare it to the technique TypeScript uses to provide dev tools. There is a compiler `tsc` which you run on the command-line to convert `*.ts` to `*.js`, then there is a `TSServer` which is a node API that responds to requests from text editors. The `TSServer` is what provides all the JavaScript and TypeScript experience, and it has most of the compiler's code inside it.
+
+For Svelte, we have the Svelte compiler, and now we have the [`svelte-language-server`](https://github.com/sveltejs/language-tools/tree/master/packages/language-server#svelte-language-server) which responds to text editor calls via the [Language Server Protocol standard](https://microsoft.github.io//language-server-protocol/overviews/lsp/overview/). First class TypeScript support means that _both_ of these two systems do good job of handling TypeScript code.
+
+The Svelte compiler support for TypeScript is handled by [Christian Kaisermann](https://github.com/kaisermann)'s [`svelte-preprocess`](https://github.com/sveltejs/svelte-preprocess#svelte-preprocess) which is now an official Svelte project.
+
+For the editor level, we took inspiration from [Pine's](https://github.com/octref) work in the [Vue](https://vuejs.org) ecosystem via [Vetur](https://github.com/vuejs/vetur). Vetur provides an [LSP](https://github.com/vuejs/vetur/blob/master/server), a VS Code extension and a [CLI](https://github.com/vuejs/vetur/blob/master/vti). Svelte now also has an [LSP](https://github.com/sveltejs/language-tools/blob/master/packages/language-server), a [VS Code extension](https://github.com/sveltejs/language-tools/blob/master/packages/svelte-vscode) and a [CLI](https://github.com/sveltejs/language-tools/blob/master/packages/svelte-check).
+
+
+#### `*.svelte` Introspection
+
+For the official Svelte VS Code extension, we built off the foundations which [James Birtles](https://github.com/UnwrittenFun) has created in [`UnwrittenFun/svelte-vscode`](https://github.com/UnwrittenFun/svelte-vscode) and [`UnwrittenFun/svelte-language-server`](https://github.com/UnwrittenFun/svelte-language-server/).
+
+[Simon H](https://github.com/dummdidumm) and [Lyu, Wei-Da](https://github.com/jasonlyu123) have worked on improving JavaScript and typeScript improvements, including integrating [@halfnelson](https://github.com/halfnelson)'s [svelte2ts](https://github.com/sveltejs/language-tools/tree/master/packages/svelte2tsx#svelte2tsx) which powers understanding the props on components in your codebase.
+
+### Try today
+
+The Svelte template has been extended with a script to convert it to a TypeScript project.
+
+```bash
+npx degit svelte/template
+node scripts/updateTypeScriptVersion.js
```

Comment:
I really think this should be turned into a `package.json` task - like `npm run typescript:init` or something. That script name is quite wordy.

Reply:
That should be discussed over at : https://github.com/sveltejs/template/pull/140

Reply:
I feel like I already mentioned it on an existing PR. I don't recognise this PR.

Reply:
I don't have _too_ strong an opinion on it, but I lean towards using the script directly. The key advantage is that it's 1 command to remove it ( `rm scripts/updateTypeScriptVersion.js`) if you're don't want TS and want to carry on in JS

if we add the script in the package JSON, then it's `rm ...` the file then edit the package manually.

Reply:
I agree with @antony, I think a `npm run ts:init` would be more comfortable, but I do understand @orta 's considerations about it.
Perhaps we could have some `svelte-cli` or similar package that would allow us to do things like `npx svelte-cli ts:init` and also `npx svelte-cli ts:check` and so. I thinks it's a clean solution.
On the other hand, I think that we should also support something like `npx degit sveltejs/template#ts my-svelte-ts-project` to start a new project with typescript support. That would defintely be the easiest way to get started without having to mess with the conf files.

Reply:
I think at this point we should probably have an `npm init svelte` command* with an interactive 'TypeScript? y/n' prompt (and possibly others). But that's a larger conversation — for now let's roll with `node scripts/setupTypeScript.js`

*this would require us to have an npm package called `create-svelte`, which [already exists](https://www.npmjs.com/package/create-svelte), so we'd need to see if the owner would consider transferring it. or we could have `@sveltejs/create` which you'd use as `npm init @sveltejs`